packages/react-components/src/components/Tooltip/Tooltip.tsx

```diff
@@ -0,0 +1,272 @@
+  arrowOffsetX?: number;
+  triggerRenderer: () => React.ReactNode;
+  referenceElement?: VirtualElement;
+  onOpen?: () => void;
+  onClose?: () => void;
+}
+
+const sleep = (milliseconds: number) => {
+  return new Promise((resolve) => setTimeout(resolve, milliseconds));
+};
+
+const baseClass = 'lc-tooltip';
+
+export const Tooltip: React.FC<ITooltipProps> = (props) => {
+  const {
+    triggerRenderer,
+    referenceElement,
+    children,
+    className,
+    theme,
+    placement,
+    isVisible = false,
+    isManaged = false,
+    withFadeAnimation = true,
+    transitionDuration = 200,
+    transitionDelay = 0,
+    hoverOutDelayTimeout = 100,
+    offsetMainAxis = 8,
+    triggerOnClick = false,
+    arrowOffsetY,
+    arrowOffsetX,
+    onOpen,
+    onClose,
+  } = props;
+
+  const arrowRef = React.useRef<HTMLDivElement | null>(null);
+  const [visible, setVisibility] = React.useState(isVisible);
+  const isHovered = React.useRef(false);
+
+  const {
+    x,
+    y,
+    reference,
+    floating,
+    strategy,
+    update,
+    refs,
+    placement: updatedPlacement,
+    middlewareData: { arrow: { x: arrowX, y: arrowY } = {} },
+  } = useFloating({
+    middleware: [
+      offset({ mainAxis: offsetMainAxis }),
+      arrow({ element: arrowRef }),
+      flip(),
+    ],
+    placement: placement,
+  });
+
+  React.useEffect(() => {
+    referenceElement && reference(referenceElement);
+  }, [reference, referenceElement]);
+
+  React.useEffect(() => {
+    setVisibility(isVisible);
+  }, [isVisible]);
+
+  React.useEffect(() => {
+    document.addEventListener('keydown', handleHideOnEscape);
+    return () => {
+      document.removeEventListener('keydown', handleHideOnEscape);
+    };
+  }, []);
+
+  React.useEffect(() => {
+    if (!refs.reference.current || !refs.floating.current) {
+      return;
+    }
+
+    // Only call this when the floating element is rendered
+    return autoUpdate(refs.reference.current, refs.floating.current, update);
+  }, [refs.reference, refs.floating, update, updatedPlacement]);
+
+  const handleMouseLeave = () => {
+    if (triggerOnClick || isManaged) return;
+    isHovered.current = false;
+    void sleep(hoverOutDelayTimeout).then(() => {
+      if (!isHovered.current) {
+        setVisibility(false);
+      }
+    });
+  };
+
+  const handleOpen = () => {
+    if (onOpen) onOpen();
+    setVisibility(true);
+  };
+
+  const handleClose = () => {
+    if (onClose) onClose();
+    setVisibility(false);
+  };
+
+  const handleMouseEnter = () => {
+    if (triggerOnClick || isManaged) return;
+    isHovered.current = true;
+    setVisibility(true);
+  };
+
+  const handleHideOnEscape = (event: KeyboardEvent) => {
+    if (event.key === 'Escape') {
+      handleClose();
+    }
+  };
+
+  const handleClick = () => {
+    if (isManaged) return;
+    if (visible) {
+      handleClose();
+    } else {
+      handleOpen();
+    }
+  };
+
+  const handleCloseOnClick = () => {
+    handleClose();
```

Comment:
we shouldnt allow closing the tooltip if it is managed
```suggestion
    if (isManaged) return;
    handleClose();
```

Reply:
Ok, after retesting the changes, I see that this one was not the best advice and needs to be reversed, apologies 🙇‍♂️ 

Mainly because in the case of the managed components we still need information about internal close button clicks. Instead of blocking the on close handler, we should only prevent changing the `visible` state directly.

So it would be best to wrap `setVisibility` calls with `if(!isManaged)` in `handleOpen` and `handleClose` 🙂

Reply:
I think we should allow this - the user can decide if he wants to show the 'x' icon or listen for the ESC key that allows customers to break whatever flow he is in.

Reply:
@MichalPaszowski exactly, you are totally right 🙂